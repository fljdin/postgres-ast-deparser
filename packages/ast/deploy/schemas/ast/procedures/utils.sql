-- Deploy schemas/ast/procedures/utils to pg

-- requires: schemas/ast/schema

BEGIN;


COMMIT;
