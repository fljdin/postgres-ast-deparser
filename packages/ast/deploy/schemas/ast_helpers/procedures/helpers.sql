-- Deploy schemas/ast_helpers/procedures/helpers to pg

-- requires: schemas/ast_helpers/schema
-- requires: schemas/ast/procedures/types 

BEGIN;


CREATE FUNCTION ast_helpers.coalesce (field text, value text default '')
    RETURNS jsonb
    AS $$
DECLARE
    result jsonb = ast.func_call(
      v_funcname := to_jsonb(ARRAY[ast.string('coalesce')]),
      v_args := to_jsonb(ARRAY[ ast.string(''), ast.a_const(ast.string('')) ])
    );
BEGIN
	result = jsonb_set(result, '{FuncCall, args, 0, String, str}', to_jsonb(field));
	result = jsonb_set(result, '{FuncCall, args, 1, A_Const, String, str}', to_jsonb(value));
	RETURN result;
END;
$$
LANGUAGE 'plpgsql'
IMMUTABLE;

CREATE FUNCTION ast_helpers.coalesce (field jsonb, value text default '')
    RETURNS jsonb
    AS $$
DECLARE
    result jsonb = ast.func_call(
      v_funcname := to_jsonb(ARRAY[ast.string('coalesce')]),
      v_args := to_jsonb(ARRAY[ ast.string(''), ast.a_const(ast.string('')) ])
    );
BEGIN
	result = jsonb_set(result, '{FuncCall, args, 0}', field);
	result = jsonb_set(result, '{FuncCall, args, 1, A_Const, String, str}', to_jsonb(value));
	RETURN result;
END;
$$
LANGUAGE 'plpgsql'
IMMUTABLE;

CREATE FUNCTION ast_helpers.tsvectorw (input jsonb, weight text='A')
    RETURNS jsonb
    AS $$
DECLARE
    result jsonb = ast.func_call(
      v_funcname := to_jsonb(ARRAY[ast.string('setweight')]),
      v_args := to_jsonb(ARRAY[input, ast.a_const(ast.string(weight))])
    );
BEGIN
	RETURN result;
END;
$$
LANGUAGE 'plpgsql'
IMMUTABLE;

CREATE FUNCTION ast_helpers.tsvector (input jsonb)
    RETURNS jsonb
    AS $$
DECLARE
    result jsonb = ast.func_call(
      v_funcname := to_jsonb(ARRAY[ast.string('to_tsvector')]),
      v_args := to_jsonb(ARRAY[input])
    );
BEGIN
	RETURN result;
END;
$$
LANGUAGE 'plpgsql'
IMMUTABLE;

CREATE FUNCTION ast_helpers.simple_param (
  name text,
  type text
)
    RETURNS jsonb
    AS $$
BEGIN
	RETURN ast.function_parameter(
      v_name := name,
      v_argType := ast.type_name( 
        v_names := to_jsonb(ARRAY[ast.string(type)])
      ),
      v_mode := 105
    );
END;
$$
LANGUAGE 'plpgsql'
IMMUTABLE;

CREATE FUNCTION ast_helpers.simple_param (
  name text,
  type text,
  default_value text
)
    RETURNS jsonb
    AS $$
BEGIN
	RETURN ast.function_parameter(
      v_name := name,
      v_argType := ast.type_name( 
        v_names := to_jsonb(ARRAY[ast.string(type)])
      ),
      v_mode := 105,
      v_defexpr := ast.string(default_value)
    );
END;
$$
LANGUAGE 'plpgsql'
IMMUTABLE;

CREATE FUNCTION ast_helpers.simple_param (
  name text,
  type text,
  default_value jsonb
)
    RETURNS jsonb
    AS $$
BEGIN
	RETURN ast.function_parameter(
      v_name := name,
      v_argType := ast.type_name( 
        v_names := to_jsonb(ARRAY[ast.string(type)])
      ),
      v_mode := 105,
      v_defexpr := default_value
    );
END;
$$
LANGUAGE 'plpgsql'
IMMUTABLE;

CREATE FUNCTION ast_helpers.tsvector (lang text, input jsonb)
    RETURNS jsonb
    AS $$
DECLARE
    result jsonb = ast.func_call(
      v_funcname := to_jsonb(ARRAY[ast.string('to_tsvector')]),
      v_args := to_jsonb(ARRAY[ast.a_const(ast.string(lang)), input])
    );
BEGIN
	RETURN result;
END;
$$
LANGUAGE 'plpgsql'
IMMUTABLE;

CREATE FUNCTION ast_helpers.a_expr_distinct_tg_field (field text)
    RETURNS jsonb
    AS $$
BEGIN
	RETURN ast.a_expr(v_kind := 3, 
        v_lexpr := ast.column_ref(
          to_jsonb(ARRAY[ ast.string('old'),ast.string(field) ])
        ),
        v_name := to_jsonb(ARRAY[ast.string('=')]),
        v_rexpr := ast.column_ref(
          to_jsonb(ARRAY[ ast.string('new'),ast.string(field) ])
        ) 
    );
END;
$$
LANGUAGE 'plpgsql'
IMMUTABLE;

CREATE FUNCTION ast_helpers.tsvector_index (fields jsonb)
    RETURNS jsonb
    AS $$
DECLARE
  results jsonb[];
  result jsonb;
  r jsonb;
	i int;
BEGIN
  FOR r IN (select jsonb_array_elements(fields))
    LOOP
      -- TODO maybe we get pg_catalog on some machines from get_current_ts_config and don't need to add it?
      IF ( (r->'lang') IS NULL) THEN
        r = jsonb_set(r, '{lang}', to_jsonb(get_current_ts_config()) );
        -- r = jsonb_set(r, '{lang}', to_jsonb('pg_catalog' || '.' || get_current_ts_config()) );
      END IF;

     -- TODO handle simple/english
      IF ( r->'array' = to_jsonb(true)) THEN
        -- handle array
        results = array_append(results, ast_helpers.tsvectorw( ast_helpers.tsvector(r->>'lang',
          -- start the string
          ast_helpers.coalesce(ast.func_call(
            v_funcname := to_jsonb(ARRAY[ast.string('array_to_string')]),
            v_args := to_jsonb(ARRAY[
              -- type cast null to text[] array
              ast.type_cast(
                v_arg := ast.string(r->>'field'),
                v_typeName := ast.type_name( 
                    v_names := to_jsonb(ARRAY[ast.string(r->>'type')]),
                    v_arrayBounds := to_jsonb(ARRAY[ast.integer(-1)])
                )
              ),
              ast.a_const(ast.string(' '))]
            )
          ))
        -- end array to string function call here
      ) , r->>'weight') );
      ELSE
        IF ( (r->'lang') IS NOT NULL) THEN
          results = array_append(results, ast_helpers.tsvectorw( ast_helpers.tsvector(r->>'lang', ast_helpers.coalesce(r->>'field')) , r->>'weight') );
        ELSE
          -- get_current_ts_config() returns 'english', we'd need to add pg_catalog on there...
          results = array_append(results, ast_helpers.tsvectorw( ast_helpers.tsvector(ast_helpers.coalesce(r->>'field')) , r->>'weight') );
        END IF;
      END IF;
    END LOOP;

  -- create the expressions
  FOR i IN SELECT * FROM generate_subscripts(results, 1) g(i)
    LOOP
      IF (i = 1) THEN
        result = results[i];
      ELSE
        result = ast.a_expr(
          v_kind := 0,
          v_lexpr := results[i], 
          v_name := to_jsonb(ARRAY[ast.string('||')]),
          v_rexpr := result );
      END IF;
    END LOOP;

	RETURN result;
END;
$$
LANGUAGE 'plpgsql'
IMMUTABLE;

CREATE FUNCTION ast_helpers.create_trigger_with_fields (
  trigger_name text,
  schema_name text,
  table_name text,
  trigger_fn_schema text,
  trigger_fn_name text,
  fields text[],
  timing int default 2,
  events int default 4 | 16
)
    RETURNS jsonb
    AS $$
DECLARE
  results jsonb[];
  result jsonb;
  whenClause jsonb;
  r jsonb;
	i int;
  field text;
BEGIN

  FOR i IN SELECT * FROM generate_subscripts(fields, 1) g(i)
    LOOP
      field = fields[i];
      IF (i = 1) THEN
        whenClause = ast_helpers.a_expr_distinct_tg_field(field);
      ELSE
        whenClause = ast.bool_expr( 
          v_boolop := 1, 
          v_args := to_jsonb(ARRAY[ast_helpers.a_expr_distinct_tg_field(field), whenClause])
        );
      END IF;
    END LOOP;

  result = ast.create_trig_stmt(
    v_trigname := trigger_name,
    v_relation := ast.range_var(schema_name, table_name, true, 'p'),
    v_funcname := to_jsonb(ARRAY[ ast.string(trigger_fn_schema),ast.string(trigger_fn_name) ]),
    v_row := true,
    v_timing := timing,
    v_events := events,
    v_whenClause := whenClause
  );


	RETURN result;
END;
$$
LANGUAGE 'plpgsql'
IMMUTABLE;


CREATE FUNCTION ast_helpers.create_function (
  schema text,
  name text,
  type text,
  parameters jsonb,
  body text,
  vol text,
  lan text,
  sec int
)
    RETURNS jsonb
    AS $$
DECLARE
  ast jsonb;
BEGIN

  select * FROM ast.create_function_stmt(
    v_funcname := to_jsonb(ARRAY[ ast.string(schema),ast.string(name) ]),
    v_parameters := parameters,
    v_returnType := ast.type_name( 
        v_names := to_jsonb(ARRAY[ast.string(type)])
    ),
    v_options := to_jsonb(ARRAY[
      ast.def_elem(
        'as',
        to_jsonb(ARRAY[ast.string(body)])
      ),
      ast.def_elem(
        'volatility',
        ast.string(vol)
      ),
      ast.def_elem(
        'language',
        ast.string(lan)
      ),
      ast.def_elem(
        'security',
        ast.integer(sec)
      )
    ]::jsonb[])
  ) INTO ast;

  RETURN ast;

END;
$$
LANGUAGE 'plpgsql'
IMMUTABLE;

CREATE FUNCTION ast_helpers.create_policy (
  name text,
  vschema text,
  vtable text,
  vrole text,
  qual jsonb,
  cmd text,
  with_check jsonb,
  permissive boolean

)
    RETURNS jsonb
    AS $$
DECLARE
  ast jsonb;
BEGIN
  select * FROM ast.create_policy_stmt(
    v_policy_name := name,
    v_table := ast.range_var(v_schemaname := vschema, v_relname := vtable, v_inh := true, v_relpersistence := 'p'),
    v_roles := to_jsonb(ARRAY[
        ast.role_spec(v_roletype:=0, v_rolename:=vrole)
    ]),
    v_qual := qual,
    v_cmd_name := cmd,
    v_with_check := with_check,
    v_permissive := permissive
  ) INTO ast;
  RETURN ast;
END;
$$
LANGUAGE 'plpgsql'
IMMUTABLE;


COMMIT;
