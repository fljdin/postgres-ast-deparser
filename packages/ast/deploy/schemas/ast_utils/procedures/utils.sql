-- Deploy schemas/ast_utils/procedures/utils to pg

-- requires: schemas/ast_utils/schema

BEGIN;

CREATE FUNCTION ast_utils.interval (
  n int
) returns text[] as $$
	select (CASE 
WHEN ( n = 2 ) THEN ARRAY[ 'month' ]
WHEN ( n = 4 ) THEN ARRAY[ 'year' ]
WHEN ( n = 6 ) THEN ARRAY[ 'year', 'month' ]
WHEN ( n = 8 ) THEN ARRAY[ 'day' ]
WHEN ( n = 1024 ) THEN ARRAY[ 'hour' ]
WHEN ( n = 1032 ) THEN ARRAY[ 'day', 'hour' ]
WHEN ( n = 2048 ) THEN ARRAY[ 'minute' ]
WHEN ( n = 3072 ) THEN ARRAY[ 'hour', 'minute' ]
WHEN ( n = 3080 ) THEN ARRAY[ 'day', 'minute' ]
WHEN ( n = 4096 ) THEN ARRAY[ 'second' ]
WHEN ( n = 6144 ) THEN ARRAY[ 'minute', 'second' ]
WHEN ( n = 7168 ) THEN ARRAY[ 'hour', 'second' ]
WHEN ( n = 7176 ) THEN ARRAY[ 'day', 'second' ]
WHEN ( n = 32767 ) THEN ARRAY[]::text[]
END);
$$  
LANGUAGE 'sql' IMMUTABLE;

CREATE FUNCTION ast_utils.reserved (
  str text
) returns boolean as $$
	select exists( select 1 from pg_get_keywords() where catcode = 'R' AND word=str  );
$$  
LANGUAGE 'sql' SECURITY DEFINER;

CREATE FUNCTION ast_utils.objtypes ()
returns text[] as $$
	select ARRAY[ 'ACCESS METHOD', 'AGGREGATE', NULL, NULL, NULL, 'CAST', 'COLUMN', 'COLLATION', 'CONVERSION', 'DATABASE', NULL, NULL, 'DOMAIN', 'CONSTRAINT', NULL, 'EXTENSION', 'FOREIGN DATA WRAPPER', 'SERVER', 'FOREIGN TABLE', 'FUNCTION', 'INDEX', 'LANGUAGE', 'LARGE OBJECT', 'MATERIALIZED VIEW', 'OPERATOR CLASS', 'OPERATOR', 'OPERATOR FAMILY', 'POLICY', NULL, NULL, 'ROLE', 'RULE', 'SCHEMA', 'SEQUENCE', NULL, 'STATISTICS', 'CONSTRAINT', 'TABLE', 'TABLESPACE', 'TRANSFORM', 'TRIGGER', 'TEXT SEARCH CONFIGURATION', 'TEXT SEARCH DICTIONARY', 'TEXT SEARCH PARSER', 'TEXT SEARCH TEMPLATE', 'TYPE', NULL, 'VIEW' ]::text[];
$$  
LANGUAGE 'sql' IMMUTABLE;

CREATE FUNCTION ast_utils.constrainttype_idxs (typ text)
returns int as $$
  select (CASE
WHEN (typ = 'CONSTR_NULL') THEN 0
WHEN (typ = 'CONSTR_NOTNULL') THEN 1
WHEN (typ = 'CONSTR_DEFAULT') THEN 2
WHEN (typ = 'CONSTR_IDENTITY') THEN 3
WHEN (typ = 'CONSTR_CHECK') THEN 4
WHEN (typ = 'CONSTR_PRIMARY') THEN 5
WHEN (typ = 'CONSTR_UNIQUE') THEN 6
WHEN (typ = 'CONSTR_EXCLUSION') THEN 7
WHEN (typ = 'CONSTR_FOREIGN') THEN 8
WHEN (typ = 'CONSTR_ATTR_DEFERRABLE') THEN 9
WHEN (typ = 'CONSTR_ATTR_NOT_DEFERRABLE') THEN 10
WHEN (typ = 'CONSTR_ATTR_DEFERRED') THEN 11
WHEN (typ = 'CONSTR_ATTR_IMMEDIATE') THEN 12
END);
$$
LANGUAGE 'sql' IMMUTABLE;

CREATE FUNCTION ast_utils.constrainttypes (contype int)
returns text as $$
  select (CASE
WHEN (contype =  0 ) THEN 'NULL'
WHEN (contype =  1 ) THEN 'NOT NULL'
WHEN (contype =  2 ) THEN 'DEFAULT'
WHEN (contype =  4 ) THEN 'CHECK'
WHEN (contype =  5 ) THEN 'PRIMARY KEY'
WHEN (contype =  6 ) THEN 'UNIQUE'
WHEN (contype =  7 ) THEN 'EXCLUDE'
WHEN (contype =  8 ) THEN 'REFERENCES'
END);
$$
LANGUAGE 'sql' IMMUTABLE;

CREATE FUNCTION ast_utils.objtypes_idxs (typ text)
returns int as $$
	select (CASE
WHEN (typ = 'OBJECT_ACCESS_METHOD') THEN 0
WHEN (typ = 'OBJECT_AGGREGATE') THEN 1
WHEN (typ = 'OBJECT_AMOP') THEN 2
WHEN (typ = 'OBJECT_AMPROC') THEN 3
WHEN (typ = 'OBJECT_ATTRIBUTE') THEN 4
WHEN (typ = 'OBJECT_CAST') THEN 5
WHEN (typ = 'OBJECT_COLUMN') THEN 6
WHEN (typ = 'OBJECT_COLLATION') THEN 7
WHEN (typ = 'OBJECT_CONVERSION') THEN 8
WHEN (typ = 'OBJECT_DATABASE') THEN 9
WHEN (typ = 'OBJECT_DEFAULT') THEN 10
WHEN (typ = 'OBJECT_DEFACL') THEN 11
WHEN (typ = 'OBJECT_DOMAIN') THEN 12
WHEN (typ = 'OBJECT_DOMCONSTRAINT') THEN 13
WHEN (typ = 'OBJECT_EVENT_TRIGGER') THEN 14
WHEN (typ = 'OBJECT_EXTENSION') THEN 15
WHEN (typ = 'OBJECT_FDW') THEN 16
WHEN (typ = 'OBJECT_FOREIGN_SERVER') THEN 17
WHEN (typ = 'OBJECT_FOREIGN_TABLE') THEN 18
WHEN (typ = 'OBJECT_FUNCTION') THEN 19
WHEN (typ = 'OBJECT_INDEX') THEN 20
WHEN (typ = 'OBJECT_LANGUAGE') THEN 21
WHEN (typ = 'OBJECT_LARGEOBJECT') THEN 22
WHEN (typ = 'OBJECT_MATVIEW') THEN 23
WHEN (typ = 'OBJECT_OPCLASS') THEN 24
WHEN (typ = 'OBJECT_OPERATOR') THEN 25
WHEN (typ = 'OBJECT_OPFAMILY') THEN 26
WHEN (typ = 'OBJECT_POLICY') THEN 27
WHEN (typ = 'OBJECT_PUBLICATION') THEN 28
WHEN (typ = 'OBJECT_PUBLICATION_REL') THEN 29
WHEN (typ = 'OBJECT_ROLE') THEN 30
WHEN (typ = 'OBJECT_RULE') THEN 31
WHEN (typ = 'OBJECT_SCHEMA') THEN 32
WHEN (typ = 'OBJECT_SEQUENCE') THEN 33
WHEN (typ = 'OBJECT_SUBSCRIPTION') THEN 34
WHEN (typ = 'OBJECT_STATISTIC_EXT') THEN 35
WHEN (typ = 'OBJECT_TABCONSTRAINT') THEN 36
WHEN (typ = 'OBJECT_TABLE') THEN 37
WHEN (typ = 'OBJECT_TABLESPACE') THEN 38
WHEN (typ = 'OBJECT_TRANSFORM') THEN 39
WHEN (typ = 'OBJECT_TRIGGER') THEN 40
WHEN (typ = 'OBJECT_TSCONFIGURATION') THEN 41
WHEN (typ = 'OBJECT_TSDICTIONARY') THEN 42
WHEN (typ = 'OBJECT_TSPARSER') THEN 43
WHEN (typ = 'OBJECT_TSTEMPLATE') THEN 44
WHEN (typ = 'OBJECT_TYPE') THEN 45
WHEN (typ = 'OBJECT_USER_MAPPING') THEN 46
WHEN (typ = 'OBJECT_VIEW') THEN 47
END);
$$  
LANGUAGE 'sql' IMMUTABLE;

CREATE FUNCTION ast_utils.getgrantobject (node jsonb)
returns text as $$
DECLARE
  objtype int;
  targtype int;
BEGIN 

  objtype = (node->'objtype')::int;
  IF (node->'targtype') IS NOT NULL THEN
    targtype = (node->'targtype')::int;
  END IF;

  IF (objtype = 0) THEN
    RETURN 'COLUMN';
  ELSIF (objtype = 1) THEN
    IF (targtype = 1) THEN 
      RETURN 'ALL TABLES IN SCHEMA';
    ELSIF (targtype = 2) THEN 
      RETURN 'TABLES';
    END IF;
    -- TODO could be a view
    RETURN 'TABLE';
  ELSIF (objtype = 2) THEN
    RETURN 'SEQUENCE';
  ELSIF (objtype = 3) THEN
    RETURN 'DATABASE';
  ELSIF (objtype = 4) THEN
    RETURN 'DOMAIN';
  ELSIF (objtype = 5) THEN
    RETURN 'FOREIGN DATA WRAPPER';
  ELSIF (objtype = 6) THEN
    RETURN 'FOREIGN SERVER';
  ELSIF (objtype = 7) THEN
    IF (targtype = 1) THEN 
      RETURN 'ALL FUNCTIONS IN SCHEMA';
    ELSIF (targtype = 2) THEN 
      RETURN 'FUNCTIONS';
    END IF;
    RETURN 'FUNCTION';
  ELSIF (objtype = 8) THEN
    RETURN 'LANGUAGE';
  ELSIF (objtype = 9) THEN
    RETURN 'LARGE OBJECT';
  ELSIF (objtype = 10) THEN
    RETURN 'SCHEMA';
  ELSIF (objtype = 11) THEN
    RETURN 'TABLESPACE';
  ELSIF (objtype = 12) THEN
    RETURN 'TYPE';
  END IF;

  RAISE EXCEPTION 'BAD_EXPRESSION %', 'GrantObjectType';

END;
$$  
LANGUAGE 'plpgsql' IMMUTABLE;

COMMIT;
