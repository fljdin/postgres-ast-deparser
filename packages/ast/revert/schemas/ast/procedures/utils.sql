-- Revert schemas/ast/procedures/utils from pg

BEGIN;

DROP FUNCTION ast.utils;

COMMIT;
