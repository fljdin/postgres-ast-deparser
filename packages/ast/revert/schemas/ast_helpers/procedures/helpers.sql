-- Revert schemas/ast_helpers/procedures/helpers from pg

BEGIN;

DROP FUNCTION ast_helpers.helpers;

COMMIT;
