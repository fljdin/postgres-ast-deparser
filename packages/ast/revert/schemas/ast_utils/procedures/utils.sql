-- Revert schemas/ast_utils/procedures/utils from pg

BEGIN;

DROP FUNCTION ast_utils.utils;

COMMIT;
