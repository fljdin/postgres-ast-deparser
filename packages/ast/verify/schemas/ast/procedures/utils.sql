-- Verify schemas/ast/procedures/utils  on pg

BEGIN;

SELECT verify_function ('ast.utils');

ROLLBACK;
