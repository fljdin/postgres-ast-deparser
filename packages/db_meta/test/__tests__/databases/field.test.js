import { getConnections } from '../../utils';
import { snap } from '../../utils/snaps';
import { initDatabase } from '../../utils/helpers';

let db, dbs, custom, teardown;
const objs = {
  tables: {}
};

beforeAll(async () => {
  ({ db, teardown } = await getConnections());
  dbs = db.helper('collections_public');

  // NOTE: POTENTIALLY YOU WILL NEED TO MOVE THIS AND BELOW CODE INTO beforeEach
  // honeslty not sure why it's not casuing issues... I'm guessing it cleans up all the created tables, etc.,
  // during the rollbacks... so it works!?!
});

beforeEach(async () => {
  await db.beforeEach();
  await initDatabase({ objs, dbs });
  custom = {
    public: db.helper(objs.database1.schema_name),
    private: db.helper(objs.database1.private_schema_name)
  };
  objs.tables.customer = await dbs.insertOne('table', {
    database_id: objs.database1.id,
    schema_id: objs.database1.schemas.public.id,
    name: 'customers'
  });
});

afterEach(async () => {
  await db.afterEach();
});

afterAll(async () => {
  await teardown();
});

it('can create a field', async () => {
  await dbs.insertOne('field', {
    table_id: objs.tables.customer.id,
    name: 'id',
    type: 'uuid',
    is_required: true,
    default_value: 'uuid_generate_v4 ()'
  });

  const nameField = await dbs.insertOne('field', {
    table_id: objs.tables.customer.id,
    name: 'name',
    type: 'text'
  });

  expect(nameField).toBeTruthy();
  expect(nameField.id).toBeTruthy();

  await db.any(
    `INSERT INTO "${objs.database1.schema_name}".customers (name) VALUES ('dan'), ('jobs')`
  );
  const customers = await db.any(
    `SELECT id, name FROM "${objs.database1.schema_name}".customers`
  );
  expect(customers).toBeTruthy();
  expect(customers.length).toBe(2);
});

it('can create an array field', async () => {
  await dbs.insertOne('field', {
    table_id: objs.tables.customer.id,
    name: 'id',
    type: 'uuid',
    is_required: true,
    default_value: 'uuid_generate_v4 ()'
  });

  const nameField = await dbs.insertOne('field', {
    table_id: objs.tables.customer.id,
    name: 'name',
    type: 'text'
  });

  const tagsField = await dbs.insertOne('field', {
    table_id: objs.tables.customer.id,
    name: 'tags',
    type: 'text[]'
  });

  const catsField = await dbs.insertOne('field', {
    table_id: objs.tables.customer.id,
    name: 'cats',
    type: 'text [ ]'
  });

  expect(nameField).toBeTruthy();
  expect(nameField.id).toBeTruthy();
  snap(tagsField);
  snap(catsField);

  await db.insertOne(
    `"${objs.database1.schema_name}".customers`,
    {
      name: ['dude', 'yo', 'hi']
    },
    { name: 'text[]' }
  );

  await db.insertOne(
    `"${objs.database1.schema_name}".customers`,
    {
      name: ['james', 'c', 'christiansen']
    },
    { name: 'text[]' }
  );

  const customers = await db.any(
    `SELECT name FROM "${objs.database1.schema_name}".customers`
  );
  expect(customers).toMatchSnapshot();
  expect(customers).toBeTruthy();
  expect(customers.length).toBe(2);
});

it('can rename fields', async () => {
  await dbs.insertOne('field', {
    table_id: objs.tables.customer.id,
    name: 'id',
    type: 'uuid',
    is_required: true,
    default_value: 'uuid_generate_v4 ()'
  });

  const referral = await dbs.insertOne('field', {
    table_id: objs.tables.customer.id,
    name: 'referral',
    type: 'uuid'
  });

  expect(referral).toBeTruthy();
  expect(referral.id).toBeTruthy();

  await custom.public.insertOne('customers', {
    referral: '929cde1e-6b8b-424c-ade9-99949a0c0212'
  });

  await dbs.update(
    'field',
    {
      name: 'referral_id'
    },
    {
      id: referral.id
    }
  );

  await custom.public.insertOne('customers', {
    referral_id: '911cde1e-6b8b-424c-ade9-99949a0c0212'
  });
});

it('can change type of fields', async () => {
  await dbs.insertOne('field', {
    table_id: objs.tables.customer.id,
    name: 'id',
    type: 'uuid',
    is_required: true,
    default_value: 'uuid_generate_v4 ()'
  });

  const referral = await dbs.insertOne('field', {
    table_id: objs.tables.customer.id,
    name: 'referral',
    type: 'uuid'
  });

  expect(referral).toBeTruthy();
  expect(referral.id).toBeTruthy();

  await custom.public.insertOne('customers', {
    referral: '929cde1e-6b8b-424c-ade9-99949a0c0212'
  });

  await dbs.update(
    'field',
    {
      type: 'text'
    },
    {
      id: referral.id
    }
  );

  await custom.public.insertOne('customers', {
    referral: 'jones'
  });

  await custom.public.update(
    'customers',
    {
      referral: '929cde1e-6b8b-424c-ade9-99949a0c0201'
    },
    {
      referral: 'jones'
    }
  );

  await dbs.update(
    'field',
    {
      type: 'uuid'
    },
    {
      id: referral.id
    }
  );
});

it('can add tags to fields', async () => {
  await dbs.insertOne('field', {
    table_id: objs.tables.customer.id,
    name: 'id',
    type: 'uuid',
    is_required: true,
    default_value: 'uuid_generate_v4 ()'
  });

  const image = await dbs.insertOne('field', {
    table_id: objs.tables.customer.id,
    description: 'Customer Profile Image',
    name: 'image',
    type: 'text'
  });

  expect(image).toBeTruthy();
  expect(image.id).toBeTruthy();

  const { get_column_smart_comment: commentBefore } = await db.one(
    `
          SELECT * FROM db_utils.get_column_smart_comment(
            $1,
            $2,
            $3
          )
          `,
    [objs.database1.schema_name, 'customers', 'image']
  );

  expect(commentBefore).toMatchSnapshot();

  await custom.public.insertOne('customers', {
    image: 'path/to/my.jpg'
  });

  await dbs.update(
    'field',
    {
      smart_tags: {
        attachment: true,
        mimetypes: ['image/jpg', 'image/png']
      },
      description: 'Customer Profile Image'
    },
    {
      id: image.id
    }
  );

  const { get_column_smart_comment: comment } = await db.one(
    `
          SELECT * FROM db_utils.get_column_smart_comment(
            $1,
            $2,
            $3
          )
          `,
    [objs.database1.schema_name, 'customers', 'image']
  );

  expect(comment).toMatchSnapshot();

  await dbs.update(
    'field',
    {
      smart_tags: {},
      description: null
    },
    {
      id: image.id
    }
  );

  const { get_column_smart_comment: commentAfter } = await db.one(
    `
          SELECT * FROM db_utils.get_column_smart_comment(
            $1,
            $2,
            $3
          )
          `,
    [objs.database1.schema_name, 'customers', 'image']
  );

  expect(commentAfter == null).toBe(true);
});

it('can delete fields', async () => {
  await dbs.insertOne('field', {
    table_id: objs.tables.customer.id,
    name: 'id',
    type: 'uuid',
    is_required: true,
    default_value: 'uuid_generate_v4 ()'
  });

  await dbs.insertOne('field', {
    table_id: objs.tables.customer.id,
    name: 'name',
    type: 'text'
  });

  const referral = await dbs.insertOne('field', {
    table_id: objs.tables.customer.id,
    name: 'referral',
    type: 'uuid'
  });

  expect(referral).toBeTruthy();
  expect(referral.id).toBeTruthy();

  await custom.public.insertOne('customers', {
    referral: '929cde1e-6b8b-424c-ade9-99949a0c0212'
  });

  await dbs.delete('field', {
    id: referral.id
  });

  const fields = await dbs.select('field', ['name', 'type'], {
    table_id: objs.tables.customer.id
  });

  expect(fields).toMatchSnapshot();

  const customers = await custom.public.select('customers', ['*']);

  expect(customers[0]).toBeTruthy();
  expect(customers[0].id).toBeTruthy();
  expect(customers[0].referral).toBeFalsy();
});

describe('naming conventions', () => {
  it('field names are snake_case', async () => {
    await dbs.insertOne('field', {
      table_id: objs.tables.customer.id,
      name: 'id',
      type: 'uuid',
      is_required: true,
      default_value: 'uuid_generate_v4()'
    });

    await dbs.insertOne('field', {
      table_id: objs.tables.customer.id,
      name: 'firstName',
      type: 'text'
    });

    await dbs.insertOne('field', {
      table_id: objs.tables.customer.id,
      name: 'LastName',
      type: 'text'
    });

    await db.any(
      `INSERT INTO "${objs.database1.schema_name}".customers(first_name, last_name) VALUES ('dan', 'lynch'), ('steve', 'jobs')`
    );
    const customers = await db.any(
      `SELECT id, first_name, last_name FROM "${objs.database1.schema_name}".customers`
    );
    expect(customers).toBeTruthy();
    expect(customers.length).toBe(2);
  });
});
describe('hidden fields', () => {
  xit('field names are snake_case', async () => {
    await db.any(
      `insert into collections_public.field 
              (table_id, name, type, is_required, default_value) 
              values 
              ($1, $2, $3, $4, $5) RETURNING *`,
      [objs.tables.customer.id, 'id', 'uuid', true, 'uuid_generate_v4 ()']
    );
    await db.any(
      'insert into collections_public.field (table_id, name, type) values ($1, $2, $3)',
      [objs.tables.customer.id, 'firstName', 'text']
    );
    await db.any(
      'insert into collections_public.field (table_id, name, type) values ($1, $2, $3)',
      [objs.tables.customer.id, 'LastName', 'text']
    );
    await db.any(
      'insert into collections_public.field (table_id, name, type, is_hidden) values ($1, $2, $3, $4)',
      [objs.tables.customer.id, 'password', 'text', true]
    );

    await db.any(
      `INSERT INTO "${objs.database1.schema_name}".customers(first_name, last_name) VALUES ('dan', 'lynch'), ('steve', 'jobs')`
    );
    const customers = await db.any(
      `SELECT id, first_name, last_name FROM "${objs.database1.schema_name}".customers`
    );
    const { get_column_smart_comment } = await db.one(
      `
            SELECT * FROM db_utils.get_column_smart_comment(
              $1,
              $2,
              $3
            )
            `,
      [objs.database1.schema_name, 'customers', 'password']
    );
    expect(get_column_smart_comment).toEqual('@omit');
    expect(customers).toBeTruthy();
    expect(customers.length).toBe(2);
  });
});
