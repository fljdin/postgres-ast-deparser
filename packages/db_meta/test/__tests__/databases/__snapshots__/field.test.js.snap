// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can add tags to fields 1`] = `"Customer Profile Image"`;

exports[`can add tags to fields 2`] = `
"@mimetypes image/jpg
@mimetypes image/png
@attachment
Customer Profile Image"
`;

exports[`can create an array field 1`] = `
Object {
  "chk": null,
  "chk_expr": null,
  "created_at": "[DATE]",
  "database_id": "[ID]",
  "default_value": null,
  "description": null,
  "field_order": 2,
  "id": "[ID]",
  "is_hidden": false,
  "is_required": false,
  "max": null,
  "min": null,
  "name": "tags",
  "regexp": null,
  "smart_tags": null,
  "table_id": "[ID]",
  "type": "text[]",
  "updated_at": "[DATE]",
}
`;

exports[`can create an array field 2`] = `
Object {
  "chk": null,
  "chk_expr": null,
  "created_at": "[DATE]",
  "database_id": "[ID]",
  "default_value": null,
  "description": null,
  "field_order": 3,
  "id": "[ID]",
  "is_hidden": false,
  "is_required": false,
  "max": null,
  "min": null,
  "name": "cats",
  "regexp": null,
  "smart_tags": null,
  "table_id": "[ID]",
  "type": "text [ ]",
  "updated_at": "[DATE]",
}
`;

exports[`can create an array field 3`] = `
Array [
  Object {
    "name": "{dude,yo,hi}",
  },
  Object {
    "name": "{james,c,christiansen}",
  },
]
`;

exports[`can delete fields 1`] = `
Array [
  Object {
    "name": "name",
    "type": "text",
  },
  Object {
    "name": "id",
    "type": "uuid",
  },
]
`;
